File: solid_diff/container.py
"""Parser for the SolidWorks 2015+ chunked container format.

Python port of the modern-format parser in openswx (MIT,
https://github.com/schwitters/openswx — see vendor/openswx/libopenswx/src/
internal/modern_parser.cc for the reference implementation and format notes).

Container layout:
  - ROL key: 1 byte at file offset 7 — rotate-left decodes all stream names.
  - Chunks are located by scanning for the 6-byte marker 14 00 06 00 08 00;
    a chunk starts 4 bytes before the marker (si = marker_pos - 4):

      si+0x00  val_a          u32   file-specific tag
      si+0x04  14 00 06 00 08 00    marker
      si+0x0a  section_type   u8    0xDF=TOC, 0xFD=data, 0x1C=mini
      si+0x0b  suffix         3B    file-specific
      si+0x0e  f1             u32   >= 65536 -> inline chunk with data
      si+0x12  csz            u32   compressed size
      si+0x16  usz            u32   uncompressed size
      si+0x1a  nsz            u32   stream-name length
      si+0x1e  name[nsz]            ROL-encoded UTF-8 stream name
      si+0x1e+nsz  data[csz]        raw-deflate payload (inline chunks only)

Pre-2015 OLE2 files are out of scope.
"""

from __future__ import annotations

import zlib
from dataclasses import dataclass, field

MARKER = b"\x14\x00\x06\x00\x08\x00"
CHUNK_HEADER_SIZE = 0x1E
INLINE_F1_THRESHOLD = 65536
MAX_NAME_SIZE = 512
MAX_COMPRESSED_SIZE = 64 * 1024 * 1024

OLE2_MAGIC = b"\xd0\xcf\x11\xe0"
ZIP_MAGIC = b"PK\x03\x04"


def rol_byte(b: int, shift: int) -> int:
    shift &= 7
    if shift == 0:
        return b
    return ((b << shift) | (b >> (8 - shift))) & 0xFF


def rol_decode(data: bytes, key: int) -> str:
    return bytes(rol_byte(b, key) for b in data).decode("latin-1")


def is_valid_stream_name(name: str) -> bool:
    return bool(name) and all(0x20 <= ord(c) < 0x80 for c in name)


def is_modern_swx(data: bytes) -> bool:
    if len(data) < 22:
        return False
    if data.startswith(OLE2_MAGIC) or data.startswith(ZIP_MAGIC):
        return False
    return data.find(MARKER, 0, 64) != -1


@dataclass
class Chunk:
    """One chunk record, including non-inline (reference) chunks."""

    offset: int
    section_type: int
    f1: int
    csz: int
    usz: int
    name: str
    data_offset: int | None = None  # absolute offset of deflate payload
    data: bytes | None = None  # decompressed payload (inline chunks)

    @property
    def inline(self) -> bool:
        return self.f1 >= INLINE_F1_THRESHOLD


@dataclass
class SwxFile:
    path: str
    rol_key: int
    chunks: list[Chunk] = field(default_factory=list)

    @property
    def streams(self) -> dict[str, bytes]:
        """First-wins map of stream name -> decompressed payload."""
        out: dict[str, bytes] = {}
        for c in self.chunks:
            if c.data is not None and c.name not in out:
                out[c.name] = c.data
        return out


def _u32(data: bytes, off: int) -> int:
    return int.from_bytes(data[off : off + 4], "little")


def parse(data: bytes, path: str = "<memory>") -> SwxFile:
    if not is_modern_swx(data):
        raise ValueError(f"{path}: not a modern (2015+) SolidWorks file")

    key = data[7]
    swx = SwxFile(path=path, rol_key=key)

    pos = 0
    while True:
        marker_pos = data.find(MARKER, pos)
        if marker_pos == -1:
            break
        if marker_pos < 4:
            pos = marker_pos + 1
            continue
        si = marker_pos - 4
        if si + CHUNK_HEADER_SIZE > len(data):
            pos = marker_pos + 1
            continue

        f1 = _u32(data, si + 0x0E)
        csz = _u32(data, si + 0x12)
        usz = _u32(data, si + 0x16)
        nsz = _u32(data, si + 0x1A)
        if nsz > MAX_NAME_SIZE or csz > MAX_COMPRESSED_SIZE:
            pos = marker_pos + 1
            continue

        name_start = si + CHUNK_HEADER_SIZE
        name_end = name_start + nsz
        if name_end > len(data):
            pos = marker_pos + 1
            continue

        name = rol_decode(data[name_start:name_end], key)
        if not is_valid_stream_name(name):
            pos = marker_pos + 1
            continue

        chunk = Chunk(
            offset=si,
            section_type=data[si + 0x0A],
            f1=f1,
            csz=csz,
            usz=usz,
            name=name,
        )

        if chunk.inline and csz > 0:
            data_end = name_end + csz
            if data_end <= len(data):
                chunk.data_offset = name_end
                try:
                    chunk.data = zlib.decompress(data[name_end:data_end], wbits=-15)
                except zlib.error:
                    chunk.data = None
                swx.chunks.append(chunk)
                pos = data_end
                continue
        elif chunk.inline:
            chunk.data = b""

        swx.chunks.append(chunk)
        pos = marker_pos + len(MARKER)

    return swx


def parse_file(path: str) -> SwxFile:
    with open(path, "rb") as f:
        return parse(f.read(), path)

File: rust/src/container.rs
//! SolidWorks 2015+ chunk container. See `docs/FORMAT.md` §1.
//!
//! Follows openswx's modern-format parser. Everything in this layer is
//! little-endian. Chunks are found by scanning for a 6-byte marker; each
//! carries a ROL-encoded stream name and, for "inline" chunks, a raw-deflate
//! payload.

use std::io::Read;

use crate::{Error, Result};

pub const MARKER: [u8; 6] = [0x14, 0x00, 0x06, 0x00, 0x08, 0x00];
pub const OLE2_MAGIC: [u8; 4] = [0xD0, 0xCF, 0x11, 0xE0];
pub const ZIP_MAGIC: [u8; 4] = [0x50, 0x4B, 0x03, 0x04];

const CHUNK_HEADER_SIZE: usize = 0x1E;
const INLINE_F1_THRESHOLD: u32 = 65536;
const MAX_NAME_SIZE: u32 = 512;
const MAX_COMPRESSED_SIZE: u32 = 64 * 1024 * 1024;

pub struct Stream {
    pub name: String,
    pub data: Vec<u8>,
}

pub struct SwxFile {
    pub rol_key: u8,
    pub chunks: Vec<Chunk>,
}

pub struct Chunk {
    pub offset: usize,
    pub section_type: u8,
    pub f1: u32,
    pub compressed_size: u32,
    pub uncompressed_size: u32,
    pub name: String,
    pub data: Option<Vec<u8>>,
}

impl Chunk {
    /// Inline chunks carry a deflate payload; the rest are references.
    pub fn inline(&self) -> bool {
        self.f1 >= INLINE_F1_THRESHOLD
    }
}

impl SwxFile {
    /// Decompressed streams, first occurrence of each name winning.
    pub fn streams(&self) -> Vec<Stream> {
        let mut seen = std::collections::HashSet::new();
        let mut out = Vec::new();
        for c in &self.chunks {
            if let Some(d) = &c.data {
                if seen.insert(c.name.clone()) {
                    out.push(Stream {
                        name: c.name.clone(),
                        data: d.clone(),
                    });
                }
            }
        }
        out
    }
}

/// Rotate one byte left by `shift & 7` bits.
fn rol_byte(b: u8, shift: u8) -> u8 {
    b.rotate_left((shift & 7) as u32)
}

/// Decode a ROL-obfuscated stream name (bytes are latin-1).
fn rol_decode(data: &[u8], key: u8) -> String {
    data.iter().map(|&b| rol_byte(b, key) as char).collect()
}

fn is_valid_stream_name(name: &str) -> bool {
    !name.is_empty() && name.chars().all(|c| ('\u{20}'..'\u{80}').contains(&c))
}

fn u32_at(data: &[u8], off: usize) -> u32 {
    u32::from_le_bytes([data[off], data[off + 1], data[off + 2], data[off + 3]])
}

fn find(haystack: &[u8], needle: &[u8], from: usize) -> Option<usize> {
    if needle.is_empty() || haystack.len() < needle.len() || from > haystack.len() - needle.len() {
        return None;
    }
    (from..=haystack.len() - needle.len()).find(|&i| &haystack[i..i + needle.len()] == needle)
}

/// Raw deflate (no zlib header/trailer), the container's payload codec.
fn inflate_raw(data: &[u8]) -> Option<Vec<u8>> {
    let mut out = Vec::new();
    let mut dec = flate2::read::DeflateDecoder::new(data);
    match dec.read_to_end(&mut out) {
        Ok(_) => Some(out),
        Err(_) => None,
    }
}

/// True for SolidWorks 2015+ chunk-container files (not OLE2, not ZIP, marker
/// present within the first 64 bytes).
pub fn is_modern_swx(data: &[u8]) -> bool {
    if data.len() < 22 {
        return false;
    }
    if data.starts_with(&OLE2_MAGIC) || data.starts_with(&ZIP_MAGIC) {
        return false;
    }
    let window = &data[..data.len().min(64)];
    find(window, &MARKER, 0).is_some()
}

/// Parse the chunk container. Malformed chunks are skipped, never fatal.
pub fn parse(data: &[u8]) -> Result<SwxFile> {
    if !is_modern_swx(data) {
        return Err(Error::NotModernSldprt);
    }

    let key = data[7];
    let mut swx = SwxFile {
        rol_key: key,
        chunks: Vec::new(),
    };

    let mut pos = 0usize;
    while let Some(marker_pos) = find(data, &MARKER, pos) {
        if marker_pos < 4 {
            pos = marker_pos + 1;
            continue;
        }
        let si = marker_pos - 4;
        if si + CHUNK_HEADER_SIZE > data.len() {
            pos = marker_pos + 1;
            continue;
        }

        let f1 = u32_at(data, si + 0x0E);
        let csz = u32_at(data, si + 0x12);
        let usz = u32_at(data, si + 0x16);
        let nsz = u32_at(data, si + 0x1A);
        if nsz > MAX_NAME_SIZE || csz > MAX_COMPRESSED_SIZE {
            pos = marker_pos + 1;
            continue;
        }

        let name_start = si + CHUNK_HEADER_SIZE;
        let name_end = name_start + nsz as usize;
        if name_end > data.len() {
            pos = marker_pos + 1;
            continue;
        }

        let name = rol_decode(&data[name_start..name_end], key);
        if !is_valid_stream_name(&name) {
            pos = marker_pos + 1;
            continue;
        }

        let mut chunk = Chunk {
            offset: si,
            section_type: data[si + 0x0A],
            f1,
            compressed_size: csz,
            uncompressed_size: usz,
            name,
            data: None,
        };

        if chunk.inline() && csz > 0 {
            let data_end = name_end + csz as usize;
            if data_end <= data.len() {
                chunk.data = inflate_raw(&data[name_end..data_end]);
                swx.chunks.push(chunk);
                pos = data_end;
                continue;
            }
        } else if chunk.inline() {
            chunk.data = Some(Vec::new());
        }

        swx.chunks.push(chunk);
        pos = marker_pos + MARKER.len();
    }

    Ok(swx)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn rol_is_self_inverse() {
        for key in 0u8..8 {
            for b in 0u8..=255 {
                assert_eq!(rol_byte(rol_byte(b, key), 8 - (key & 7)), b);
            }
        }
    }

    #[test]
    fn rejects_ole2_and_zip() {
        let mut ole = OLE2_MAGIC.to_vec();
        ole.extend_from_slice(&[0u8; 64]);
        assert!(!is_modern_swx(&ole));
        let mut zip = ZIP_MAGIC.to_vec();
        zip.extend_from_slice(&[0u8; 64]);
        assert!(!is_modern_swx(&zip));
        assert!(!is_modern_swx(&[0u8; 8]));
    }
}
